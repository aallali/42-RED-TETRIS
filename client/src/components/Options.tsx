/* eslint-disable react-hooks/exhaustive-deps */
import React, { useEffect, useState } from "react"

// Hooks
import { useAppDispatch, useAppSelector } from "../app/hooks";

// assets
import audioOn from "../assets/images/mute.png"
import audioOff from "../assets/images/sound.png"
import locked from "../assets/images/lock.png"
import unlocked from "../assets/images/unlocked.png"
import plus from "../assets/images/plus.png"
import minus from "../assets/images/minus.png"
import "../pages/playboard.css"
// Actions
import { UPDATE_GAME_MODE, UPDATE_ROOM_SIZE } from "../app/actions"
// Selectors
import { isAdmin as ifAdmin } from "../reducers/player.reducer"
import { getGameMode } from "../reducers/game.reducer";


const Options: React.FC = () => {
	const dispatch = useAppDispatch()
	const gameMode = useAppSelector<string>(getGameMode)
	const isAdminS = useAppSelector<boolean>(ifAdmin)
	const [soundOn, toggleSound] = useState<boolean>(true)
	const [isLocked, toggleLock] = useState<boolean>(gameMode === "solo" ? true : false)
	const [roomSize, setRoomSize] = useState<number>(gameMode === "solo" ? 1 : 10)
	useEffect(() => {
		if (isLocked)
			dispatch(UPDATE_GAME_MODE("solo"))
		else
			dispatch(UPDATE_GAME_MODE("multiplayer"))
	}, [isLocked])

	useEffect(() => {
		if (roomSize > 1 && roomSize <= 10)
			dispatch(UPDATE_ROOM_SIZE(roomSize))

	}, [roomSize])
	return (
		<div className="font-bold backdrop-filter backdrop-blur-xs" >
			<p >
				<button className="inline-flex items-center justify-center w-9 h-9 m-1 text-indigo-100 transition-colors duration-150 rounded-lg focus:shadow-outline bg-yellow-300 hover:bg-yellow-500"
					data-testid="toggleAudio"
					onClick={() => toggleSound(prev => !prev)}>
					<img className="h-5" src={!soundOn ? audioOn : audioOff} alt="sound" />
				</button>
				<i className="text-sm ">: {soundOn ? "Sound On" : "Sound Off"}</i>
			</p>
			{isAdminS ? (
				<p>
					<button
						className="inline-flex items-center justify-center w-9 h-9 m-1 text-indigo-100 transition-colors duration-150 rounded-lg focus:shadow-outline bg-yellow-300 hover:bg-yellow-500"
						data-testid="toggleMode"
						onClick={() => toggleLock(prev => !prev)}
					>
						<img className="h-5" src={isLocked ? locked : unlocked} alt="sound" />
					</button>
					<i className="text-sm">: {isLocked ? "Locked Mode" : "Open Mode"}</i>
				</p>) : null}


			{isAdminS ? (<div className="flex flex-row">
				<div className=" flex flex-row  ml-1 mt-1 h-8 w-20 rounded-lg  relative align-middle">
					<button
						className="
						font-semibold
						bg-yellow-300
						hover:bg-yellow-500
						textblack
						h-full
						w-14
						flex rounded-l
						focus:outline-none
						cursor-pointer
						 align-middle"
						onClick={() => setRoomSize((prev) => prev > 1 ? --prev : prev)}
						data-testid="decreaseSize"
					>
						<span className="m-auto">
							<img className="h-3" src={minus} alt="minus" />
						</span>
					</button>
					{/* <input type="hidden" className="md:p-0 p-0 text-xs md:text-base  focus:outline-none text-center" readOnly name="custom-input-number" /> */}
					<div className="bg-white text-black w-14 text-xs md:text-base flex items-center justify-center cursor-default font-bold">
						<span>{roomSize}</span>
					</div>
					<button
						className="font-semibold
						bg-yellow-300
						hover:bg-yellow-500
						text-black
						h-full
						w-14 flex
						rounded-r
						focus:outline-none
						cursor-pointer"
						data-testid="increaseSize"
						onClick={() => setRoomSize((prev) => prev < 10 ? ++prev : prev)}
					>
						<span className="m-auto">
							<img className="h-3" src={plus} alt="plus" />
						</span>
					</button>
				</div>
				<div className="pt-4">
					<i className=""> : Room Size</i>
				</div>

			</div>) : null}

		</div>

	)
}


export default Options
