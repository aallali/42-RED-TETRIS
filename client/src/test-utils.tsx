// test-utils.jsx
import React, { ReactChildren } from 'react'
import { render as rtlRender } from '@testing-library/react'
import { configureStore } from '@reduxjs/toolkit'
import { Provider } from 'react-redux'
// Import your own reducer
import playerReducer from "./reducers/player.reducer";
import opponentReducer from "./reducers/opponent.reducer"
import errorReducer from "./reducers/error.reducer"
import gameReducer from "./reducers/game.reducer"
import chatReducer from './reducers/chat.reducer';

function aallaliStore(preloadedState: any) {
	return configureStore({
		reducer: {
			player: playerReducer,
			opponents: opponentReducer,
			error: errorReducer,
			game: gameReducer,
			chat: chatReducer
		},
		preloadedState
	})
}
function render(
	ui: React.ReactElement<any, string | React.JSXElementConstructor<any>>,
	{
		preloadedState,
		store = configureStore({
			reducer: {
				player: playerReducer,
				opponents: opponentReducer,
				error: errorReducer,
				game: gameReducer,
				chat: chatReducer
			},
			preloadedState
		}),
		...renderOptions
	}: any = {}
) {
	function Wrapper({ children }: { children: ReactChildren }) {
		return <Provider store={store}>{children}</Provider>

	}
	return rtlRender(ui, { wrapper: Wrapper, ...renderOptions })
}


// re-export everything
export * from '@testing-library/react'
// override render method
export { render, aallaliStore }