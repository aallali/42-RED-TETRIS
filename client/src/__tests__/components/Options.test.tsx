import renderer from "react-test-renderer";
import Options from "../../components/Options"
import { Provider } from 'react-redux';

import { aallaliStore, render, screen } from '../../test-utils'
import { fireEvent } from "@testing-library/react";


describe("_ Options Component test", () => {
	test("render test 1", () => {
		const store = aallaliStore({ game: { mode: "solo" } })
		const tree = renderer.create(<Provider store={store}>
			< Options />
		</Provider>).toJSON();
		expect(tree).toMatchSnapshot();
	});
	test("render test 2", async () => {
		render(
			< Options />, {
			preloadedState:
			{
				opponents: {
					players: [{
						name: "testOpponent",
						lost: true,
						stage: [],
						rows: 55,
						score: 55,
						level: 55,
						admin: true
					}]
				},
				game: { mode: "solo" },
				player: { isAdmin: true }
			}
		}
		)

		fireEvent.click(screen.getByTestId("toggleAudio"))
		expect(await screen.findByText(/Sound Off/i)).toBeInTheDocument()

		fireEvent.click(screen.getByTestId("toggleAudio"))
		expect(await screen.findByText(/Sound On/i)).toBeInTheDocument()

		fireEvent.click(screen.getByTestId("toggleMode"))
		expect(await screen.findByText(/open/i)).toBeInTheDocument()

		fireEvent.click(screen.getByTestId("toggleMode"))
		expect(await screen.findByText(/locked/i)).toBeInTheDocument()


		fireEvent.click(screen.getByTestId("increaseSize"))
		fireEvent.click(screen.getByTestId("increaseSize"))
		expect(await screen.findByText(/3/i)).toBeInTheDocument()

		fireEvent.click(screen.getByTestId("decreaseSize"))
		expect(await screen.findByText(/2/i)).toBeInTheDocument()

		fireEvent.click(screen.getByTestId("decreaseSize"))
		expect(await screen.findByText(/1/i)).toBeInTheDocument()

		fireEvent.click(screen.getByTestId("decreaseSize"))
		expect(await screen.findByText(/1/i)).toBeInTheDocument()

		fireEvent.click(screen.getByTestId("increaseSize")) // 2
		fireEvent.click(screen.getByTestId("increaseSize")) // 3
		fireEvent.click(screen.getByTestId("increaseSize")) // 4
		fireEvent.click(screen.getByTestId("increaseSize")) // 5
		fireEvent.click(screen.getByTestId("increaseSize")) // 6
		fireEvent.click(screen.getByTestId("increaseSize")) // 7
		fireEvent.click(screen.getByTestId("increaseSize")) // 8
		fireEvent.click(screen.getByTestId("increaseSize")) // 9
		fireEvent.click(screen.getByTestId("increaseSize")) // 10
		fireEvent.click(screen.getByTestId("increaseSize")) // should not increase to 11 after fire another click
		expect(await screen.findByText(/10/i)).toBeInTheDocument()

	});

	test("render test 3", async () => {
		render(
			< Options />, {
			preloadedState:
			{
				opponents: {
					players: [{
						name: "testOpponent",
						lost: true,
						stage: [],
						rows: 55,
						score: 55,
						level: 55,
						admin: true
					}]
				},
				game: { mode: "multiplayer" },
				player: { isAdmin: true }
			}
		}
		)

		expect(await screen.findByText(/Open Mode/i)).toBeInTheDocument()
		expect(await screen.findByText(/10/i)).toBeInTheDocument()

	});

})
