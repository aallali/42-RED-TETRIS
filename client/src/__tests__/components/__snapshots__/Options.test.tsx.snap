// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_ Options Component test render test 1 1`] = `
<div
  className="font-bold backdrop-filter backdrop-blur-xs"
>
  <p>
    <button
      className="inline-flex items-center justify-center w-9 h-9 m-1 text-indigo-100 transition-colors duration-150 rounded-lg focus:shadow-outline bg-yellow-300 hover:bg-yellow-500"
      data-testid="toggleAudio"
      onClick={[Function]}
    >
      <img
        alt="sound"
        className="h-5"
        src="sound.png"
      />
    </button>
    <i
      className="text-sm "
    >
      : 
      Sound On
    </i>
  </p>
</div>
`;
